File: 152.maximum-product-subarray.c
/*
 * @lc app=leetcode id=152 lang=c
 *
 * [152] Maximum Product Subarray
 */
int max(int x, int y)
{

    if (x > y)
    {
        return x;
    }
    else
    {
        return y;
    }
}

int min(int x, int y)
{

    if (x < y)
    {
        return x;
    }
    else
    {
        return y;
    }
}

// @lc code=start
int maxProduct(int *nums, int numsSize)
{

    int res = nums[0];
    int currentMin = 1;
    int currentMax = 1;

    // set to max int in nums, becaus its the minimum max!
    for (int i = 1; i < numsSize; i++)
    {
        if (nums[i] > res)
        {
            res = nums[i];
        }
    }

    // iterate trough numbers
    for (int i = 0; i < numsSize; i++)
    {
        // handle specal case
        if (nums[i] == 0)
        {
            currentMin = 1;
            currentMax = 1;
            continue;
        }

        // recompute current min and max
        int tempcurrentMin = fmin(fmin(nums[i], nums[i] * currentMin), nums[i] * currentMax); // create temp, becous we need it in current max calculaton
        currentMax = fmax(fmax(nums[i], nums[i] * currentMax), nums[i] * currentMin);
        currentMin = tempcurrentMin;
        res = fmax(res, currentMax);
    }

    return res;
}
// @lc code=end

File: 152.maximum-product-subarray.py
#
# @lc app=leetcode id=152 lang=python3
#
# [152] Maximum Product Subarray
#

# @lc code=start
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        solution = max(nums) #init to max 
        curMIN, curMAX = 1,1 #init to 1, because it is neutral in multiplications

        for n in nums:
            
            temp_curMAX = curMAX
            curMAX = max(n*curMAX, n*curMIN, n) #new max can be n times min or max or n
            curMIN = min(n*temp_curMAX, n*curMIN, n)

            #update result
            solution = max(solution, curMAX)

        return solution
# @lc code=end

